### ISSUE TITLE
twisted.internet.qtreactor tries to import a module which never exists

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2608]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2608|
|Type|defect|
|Created|2007-04-26 15:56:11Z|
|Branch|https://github.com/twisted/twisted/tree/qtreactor-qtstub-2608|

There is no qtreactor_stub.  It means to import twisted_qtstub.

Attachments:

* [qtreactor-import.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a94/a9412422c11252f60268edc332c9efaf8a40f6e7/42c705047fdd1b92358f369e1313a85228b196b3.patch) (1629 bytes) - added by jesstess on 2009-09-13 21:42:03Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2608 2608
type__defect defect
reporter__exarkun exarkun
priority__low low
milestone__ 
branch__branches_qtreactor_qtstub_2608 branches/qtreactor-qtstub-2608
branch_author__jesstess jesstess
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1177602971000000 1177602971000000
changetime__1261675481000000 1261675481000000
version__None None
owner__ 
cc__jesstess cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_2"></a>|@jesstess removed owner|
|-|-|

I added a copyright.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @jesstess|
|-|-|

We should add a unit test for the error path through this code.  The test can just import `twisted.internet.qtreactor` (perhaps after doing something to `sys.modules` to ensure that the `qtreactor` import fails even if by some strange circumstance `qtreactor` is installed) and assert that an `ImportError` with the right arguments is raised.

At some point we'll probably also want to do something to this code relating to Qt4 and a Qt4 reactor ([#3977](https://github.com/twisted/twisted/issues/3977)).  But since that's not done yet, it probably has no implications for this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_4"></a>|@jesstess removed owner|
|-|-|

Added a test case in twisted/test/test_application.py. I still don't have a convenient way to test changes on a Windows machine so that hasn't been done.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_5"></a>|@washort set status to `closed`|
|-|-|

Test passes on the Windows buildbot. Looks good -- I am merging it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `reopened`|
|-|-|

Yea, how about that?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to dash<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_8"></a>|@jesstess set status to `closed`|
|-|-|

(In [27838]) Apply qtreactor-import.patch

Author: jesstess
Reviewers: exarkun, washort
Fixes: [#2608](https://github.com/twisted/twisted/issues/2608)

twisted.internet.qtreactor no longer tries to import the non-existent
qtreactor_stub and instead imports twisted_qtstub.

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|TimAllen set status to `reopened`|
|-|-|

This commit seems to have broken Python2.4:

 * [ubuntu-py2.4-select](http://buildbot.twistedmatrix.com/builders/ubuntu-py2.4-select/builds/48/steps/trial/logs/problems)
 * [debian64-py2.4-select](http://buildbot.twistedmatrix.com/builders/debian64-py2.4-select/builds/48/steps/trial/logs/problems)
 * [debian-py2.4-gtk2](http://buildbot.twistedmatrix.com/builders/debian-py2.4-gtk2/builds/48/steps/gtk2/logs/problems)

All the above logs have a single test failure; the test added in this patch:

```
[ERROR]: twisted.test.test_application.PluggableReactorTestCase.test_importQtreactor

Traceback (most recent call last):
  File "/var/lib/buildbot/twisted-bigdogvm1/ubuntu-py2.4-select/Twisted/twisted/test/test_application.py", line 883, in test_importQtreactor
    self.assertEquals(e.message, errorMessage)
exceptions.AttributeError: ImportError instance has no attribute 'message'
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_10"></a>|@radix commented|
|-|-|

(In [27847]) Revert r27838: Incompatibility with Python 2.4

The error as reported by all of the Python 2.4 buildslaves is as follows:

[ERROR]: twisted.test.test_application.PluggableReactorTestCase.test_importQtreactor
Traceback (most recent call last):
  File "/var/lib/buildbot/twisted-bigdogvm1/ubuntu-py2.4-select/Twisted/twisted/test/test_application.py", line 883, in test_importQtreactor
    self.assertEquals(e.message, errorMessage)
exceptions.AttributeError: ImportError instance has no attribute 'message'

Reopens: [#2608](https://github.com/twisted/twisted/issues/2608)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve commented|
|-|-|

The test should probably also uses `assertRaises`, unless I'm missing something.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_12"></a>|@jesstess commented|
|-|-|

(In [27849]) Branching to 'qtreactor-qtstub-2608'

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_13"></a>|@jesstess removed owner<br>@jesstess set status to `new`|
|-|-|

Use str() instead of the message attribute on ImportError to get the error message, as message wasn't an attribute before python 2.5.

Buildbot results [here](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/qtreactor-qtstub-2608).

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|TimAllen set owner to TimAllen|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|TimAllen set owner to @jesstess|
|-|-|

Works fine now, thanks!

Two little nits:
 1. `test_importQtreactor` defines a nested function called `test_Import()` but doesn't actually call it.
 2. It seems odd that the code for testing `t.i.qtreactor` is in `t.test.test_application` rather than, say, `t.i.test.test_qtreactor`. Is there an important reason for keeping this test with the Twisted Application code?

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_16"></a>|@jesstess commented|
|-|-|

(In [27852]) Move qtreactor tests into their own file under internet.

refs [#2608](https://github.com/twisted/twisted/issues/2608)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_17"></a>|@jesstess removed owner|
|-|-|

Points by TimAllen addressed.

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|TimAllen set owner to TimAllen|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_19"></a>|TimAllen set owner to @jesstess|
|-|-|

I see you moved the pre-existing `test_qtStub()` to the new location along with the new `test_importQtreactor`. It turns out this test is broken in a rather impressive number of ways:

 1. It calls `getProcessOutputAndValue()` passing `sys.executable` as the first item in the `args` tuple, exactly what the docs for that function say *not* to do. This results in the Python binary trying to execute itself as a Python script and barfing, but...
 1. ...Python only barfs on stderr, it doesn't print anything on stdout. The test only checks the contents of stdout, and concludes that Everything's Fine. Unfortunately...
 1. ...the Python script the test tries to execute wasn't moved when the test was moved, so the test can't even find the script - and yet the test still passes.
 1. After fixing all these problems, the test now fails because of the new, more reliable import-guarding added by this ticket.

My verdict: delete the poorly-written and overly-complex `test_qtStub()`, delete `twisted/test/app_qtstub.py` - your new test is much better than the old, and deserves to replace it.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_20"></a>|@jesstess commented|
|-|-|

(In [27860]) Delete redundant/broken import test case.

refs [#2608](https://github.com/twisted/twisted/issues/2608)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_21"></a>|@jesstess commented|
|-|-|

(In [27861]) app_qtstub.py is no longer used anywhere.

refs [#2608](https://github.com/twisted/twisted/issues/2608)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|TimAllen set owner to TimAllen|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_24"></a>|TimAllen set owner to @jesstess|
|-|-|

Looks good to me! Apologies for messing you around with all these reviews, but it's a really polished patch now. Merge away!

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_25"></a>|@jesstess set status to `closed`|
|-|-|

(In [27869]) Merge qtreactor-qtstub-2608

Author: jesstess
Reviewers: exarkun, washort, TimAllen, therve, 
Fixes: [#2608](https://github.com/twisted/twisted/issues/2608)

twisted.internet.qtreactor no longer tries to import the non-existent
qtreactor_stub and instead imports twisted_qtstub.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
